lib/basedriver/capabilities.js

```diff
@@ -0,0 +1,111 @@
+import _ from 'lodash';
+import { validator } from './desired-caps';
+import { util } from 'appium-support';
+
+// Takes primary caps object and merges it into a secondary caps object.
+// (see https://www.w3.org/TR/webdriver/#dfn-merging-capabilities)
+function mergeCaps (primary = {}, secondary = {}) {
+  let result = Object.assign(primary);
+
+  for (let [name, value] of _.toPairs(secondary)) {
+    // Overwriting is not allowed. Primary and secondary must have different properties (w3 rule 4.4)
+    if (!_.isUndefined(primary[name])) {
+      throw new Error(`property ${name} should not exist on both primary and secondary`);
+    }
+    result[name] = value;
+  }
+
+  return result;
+}
+
+// Validates caps against a set of constraints
+function validateCaps (caps, constraints = {}, skipPresenceConstraint = false) {
+  let message = '';
+  if (!_.isObject(caps)) {
+    throw new Error(`must be a JSON object`);
+  }
+
+  constraints = _.cloneDeep(constraints); // Defensive copy
+
+  // Remove the 'presence' constraint if we're not checking for it
+  if (skipPresenceConstraint) {
+    constraints = Object.assign(constraints);
+    for (let key of Object.keys(constraints)) {
+      delete constraints[key].presence;
+    }
+  }
+
+  let validationErrors = validator.validate(_.pickBy(caps, util.hasValue),
+                                              constraints,
+                                              {fullMessages: false});
+                                          
+  if (validationErrors) {
```

Comment:
The `message` variable is localized to this conditional, so declare it here. Also, use an array and then `join` at the end.